flocker/node/agents/blockdevice.py

```diff
@@ -53,6 +59,125 @@ class UnattachedVolume(VolumeException):
+    u"The unique identifier of a dataset."
+)
+
+DATASET_ID = Field(
+    u"dataset_id",
+    lambda dataset_id: unicode(dataset_id),
+    u"The unique identifier of a dataset."
+)
+
+MOUNTPOINT = Field(
+    u"mountpoint",
+    lambda path: path.path,
+    u"The absolute path to the location on the node where the dataset will be "
+    u"mounted.",
+)
+
+DEVICE_PATH = Field(
+    u"block_device_path",
+    lambda path: path.path,
+    u"The absolute path to the block device file on the node where the "
+    u"dataset is attached.",
+)
+
+BLOCK_DEVICE_ID = Field(
+    u"block_device_id",
+    lambda id: unicode(id),
+    u"The unique identifier if the underlying block device."
+)
+
+BLOCK_DEVICE_SIZE = Field(
+    u"block_device_size",
+    lambda size: size,
+    u"The size of the underlying block device."
+)
+
+BLOCK_DEVICE_HOST = Field(
+    u"block_device_host",
+    lambda host: host,
+    u"The host to which the underlying block device is attached."
+)
+
+CREATE_BLOCK_DEVICE_DATASET = ActionType(
+    u"agent:blockdevice:create",
+    [DATASET, MOUNTPOINT],
+    [DEVICE_PATH, BLOCK_DEVICE_ID, DATASET_ID, BLOCK_DEVICE_SIZE,
+     BLOCK_DEVICE_HOST],
+    u"A block-device-backed dataset is being created.",
+)
+
+
+@implementer(IStateChange)
+class CreateBlockDeviceDataset(PRecord):
+    """
+    An operation to create a new dataset on a newly created volume with a newly
+    initialized filesystem.
+
+    :ivar Dataset dataset: The dataset for which to create a block device.
+    :ivar FilePath mountpoint: The path at which to mount the created device.
+    :ivar Logger logger: An Eliot ``Logger``.
+    """
+    dataset = field()
+    mountpoint = field(mandatory=True)
+
+    logger = Logger()
+
+    def run(self, deployer):
+        """
+        Create a block device, attach it to the host of the supplied
+        ``deployer``, create an ``ext4`` filesystem on the device and mount it.
+
+        Operations are performed synchronously.
+
+        See ``IStateChange.run`` for general argument and return type
+        documentation.
+
+        :returns: An already fired ``Deferred`` with result ``None``.
+        """
+        with CREATE_BLOCK_DEVICE_DATASET(
+                self.logger,
+                dataset=self.dataset, mountpoint=self.mountpoint
+        ) as action:
+            api = deployer.block_device_api
+            volume = api.create_volume(
+                dataset_id=UUID(self.dataset.dataset_id),
+                size=self.dataset.maximum_size,
+            )
+
+            # This will be factored into a separate IStateChange to support the
+            # case where the volume exists but is not attached.  That object
+            # will be used by this one to perform this work.  FLOC-1575
+            volume = api.attach_volume(
+                volume.blockdevice_id, deployer.hostname
+            )
+            device = api.get_device_path(volume.blockdevice_id)
+            self.mountpoint.makedirs()
```

Comment:
If it already exists this will throw exception. Follow up issue, or add requirement to existing one.